sdlib/d/gc/arena.d

```diff
@@ -374,6 +428,29 @@ private:
 		unusedExtents.insert(e);
 	}
 
+	void shrinkAllocImpl(Extent* e, uint n, uint oldPages, uint delta) {
+		assert(mutex.isHeld(), "Mutex not held!");
+		assert(oldPages > 0 && oldPages <= PageCount,
+		       "Invalid number of old pages!");
+		assert(delta != 0, "Invalid delta!");
+		assert(n <= PageCount - oldPages, "Invalid index!");
+
+		auto hpd = e.hpd;
+		auto prevIndex = getFreeSpaceClass(hpd.longestFreeRange);
+
+		hpd.shrinkAlloc(n + oldPages - delta, delta);
+
+		auto index = getFreeSpaceClass(hpd.longestFreeRange);
+		if (index != prevIndex) {
+			heaps[prevIndex].remove(hpd);
+			filter &= ~(ulong(heaps[prevIndex].empty) << prevIndex);
```

Comment:
If the hpd was full, then this is going to fail. This bug was entirely possible because, quoting:
> Also, looking at freePagesImpl, it looks like there is a lot of overlap.

freePagesImpl handles this properly.